packages/flutter/lib/src/widgets/icon.dart

```diff
@@ -231,14 +233,31 @@ class Icon extends StatelessWidget {
   /// specified, either directly using this property or using [Directionality].
   final TextDirection? textDirection;
 
+  /// Whether to scale the size of this widget using the ambient [MediaQuery]'s [TextScaler].
+  ///
+  /// This is specially useful when you have an icon associated with a text, as
+  /// scaling the text without scaling the icon would result in a confusing
+  /// interface.
+  ///
+  /// Defaults to the nearest [IconTheme]'s
+  /// [IconThemeData.applyTextScaling].
+  final bool? applyTextScaling;
+
   @override
   Widget build(BuildContext context) {
     assert(this.textDirection != null || debugCheckHasDirectionality(context));
     final TextDirection textDirection = this.textDirection ?? Directionality.of(context);
 
     final IconThemeData iconTheme = IconTheme.of(context);
 
-    final double? iconSize = size ?? iconTheme.size;
+    final bool? applyTextScaling = this.applyTextScaling ?? iconTheme.applyTextScaling;
+
+    final double? tentativeIconSize = size ?? iconTheme.size;
+
+    final double? iconSize = switch ((tentativeIconSize, applyTextScaling)) {
```

Comment:
If `applyingTextScaling  == null`  then it's treated as if it was set to false. Then why are we allowing setting the value to null in the theme? It doesn't have a different meaning than setting the value to false in the theme right?

Reply:
Yes. I am actually following the pattern established for other properties of the theme. Take, for instance, `size`. It is also nullable, and the default value (`24.0`) is set in the `fallback()` constructor. Then, in the `Icon` widget, we simply take it as nullable and pass down to the text style.

It happens with all other properties of the theme, so I feel it's reasonable to follow it.

Reply:
Ah ok makes sense to me. But in that case I think you'll have to change the implementation of a few more methods (`==`, `merge` for example).

Reply:
`==` already considers `applyTextScaling`.

`merge` was unintentionally left out, so I am going to work on it.

Finally, there's `lerp`, but I am not sure how I should handle it. Should I scale the size within the `lerp` method? Like, for instance, lerping from `(size: 10.0)` to `(size: 10.0, applyTextScaling: true)` with a scale factor of 2.0 will yield a size of 15.0 when `t` is `0.5` or something like this... WDYT?

Reply:
I implemented both `merge` and `lerp`. For `lerp`, I looked at how it's done in other themes' `lerp` methods, and all of them simply do `t < 0.5 ? a?.property : b?.property`, so I went with this.

I am a little confused with `isConcrete`, though. I couldn't understand why `shadows` are not considered int it. It's used in two places, and one of them is in `IconTheme.of`:

```dart
  static IconThemeData of(BuildContext context) {
    final IconThemeData iconThemeData = _getInheritedIconThemeData(context).resolve(context);
    return iconThemeData.isConcrete
      ? iconThemeData
      : iconThemeData.copyWith(
        size: iconThemeData.size ?? const IconThemeData.fallback().size,
        fill: iconThemeData.fill ?? const IconThemeData.fallback().fill,
        weight: iconThemeData.weight ?? const IconThemeData.fallback().weight,
        grade: iconThemeData.grade ?? const IconThemeData.fallback().grade,
        opticalSize: iconThemeData.opticalSize ?? const IconThemeData.fallback().opticalSize,
        color: iconThemeData.color ?? const IconThemeData.fallback().color,
        opacity: iconThemeData.opacity ?? const IconThemeData.fallback().opacity,
        shadows: iconThemeData.shadows ?? const IconThemeData.fallback().shadows,
        applyTextScaling: iconThemeData.applyTextScaling ?? const IconThemeData.fallback().applyTextScaling,
      );
  }
```

As `shadows` are not included in `isConcrete`, if all properties are non-null but shadows are null, the fallback shadows won't be passed. I think this is not the intended behavior. (BTW, I implemented the shadow property in `IconTheme`, and not passing it to `isConcrete` was probably because I overlooked it. However, it was later changed to include other properties and purposely left out, as shown by the documentation)

Why is `shadows` not included? What defines what should be included in it or not? And, ultimately, should I include `applyScalingFactor` in it?

Reply:
Shadow == null means it's off (although it feels a bit strange). So the fallback value of `applyScalingFactor` shouldn't be null (being null typically means it should fall back to the default value).

Reply:
So yes I think should be taken into account by the `isConcrete` implementation.

Reply:
Done.